crates/trippy-tui/src/locale.rs

```diff
@@ -1,43 +1,125 @@
 }
 
-/// Get the current locale.
-pub fn locale() -> String {
-    rust_i18n::locale().to_string()
+/// Get the available locales.
+pub fn available_locales() -> Vec<&'static str> {
+    data()
+        .0
+        .iter()
+        .flat_map(|(_, v)| v.0.keys().map(AsRef::as_ref))
+        .unique()
+        .sorted_unstable()
+        .collect::<Vec<_>>()
 }
 
-/// Get all available locales.
-pub fn available_locales() -> Vec<&'static str> {
-    rust_i18n::available_locales!()
+/// A macro to translate an item to the current locale.
+#[macro_export]
+macro_rules! t {
+    ($key:expr) => {
+        std::borrow::Cow::Borrowed($crate::locale::__translate($key))
+    };
+    ($key:expr, $($kt:ident = $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", stringify!($kt), "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+    ($key:expr, $($kt:literal => $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", $kt, "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+}
+
+/// Translate an item to the current locale.
+///
+/// This function is public as it is used by the `t!` macro, however is not considered part of the
+/// public interface.
+#[doc(hidden)]
+pub fn __translate(item: &str) -> &str {
+    translate_locale(item, locale().load().as_str())
+}
+
+/// Translate an item to a specific locale.
+///
+/// If the item does not exists, the key is returned. Otherwise, if item does not contain the
+/// locale, the fallback locale is used. If the fallback locale does not exist, the key is
+/// returned.
+fn translate_locale<'a>(item: &'a str, locale: &'_ str) -> &'a str {
+    if let Some(key) = data().0.get(item) {
+        if let Some(value) = key.0.get(locale) {
+            value
+        } else if let Some(value) = key.0.get(&split_locale(locale)) {
+            value
+        } else if let Some(value) = key.0.get(FALLBACK_LOCALE) {
+            value
+        } else {
+            item
+        }
+    } else {
+        item
+    }
+}
+
+/// Get the locale data.
+fn data() -> &'static Data {
+    static DATA: OnceLock<Data> = OnceLock::new();
+    DATA.get_or_init(|| {
+        toml::from_str(include_str!("../locales.toml")).expect("Failed to parse locales.toml")
+    })
 }
 
-fn set_locale_inner(locale: &str) {
-    let all_locales = rust_i18n::available_locales!();
-    if all_locales.contains(&locale) {
-        rust_i18n::set_locale(locale);
+/// This is a map of a item name (i.e. `title_hops`, `awaiting_data`, etc.) to the locale `Item`.
+#[derive(Debug, Deserialize)]
+struct Data(HashMap<String, Item>);
+
+/// This is a map of locale keys (i.e. `en`, `zh`, etc.) to the translated value.
+#[derive(Debug, Deserialize)]
+struct Item(HashMap<String, String>);
+
+fn set_locale_inner(locale: String) -> String {
```

Comment:
This logic has gone missing I think; well as doing fallback at item-translation-time there is also fallback at locale-selecton-time.

Reply:
Maybe you are worried that falling back on every string selection (or item-translation-time) costs more than one time on the `set_locale()` call, and I can't deny that. Though then I'd argue locale tag splitting should also happen here, otherwise it would have the same cost burden.

Aesthetically I'm curious why this is in an "inner" function, but it's not that important.

Reply:
> Maybe you are worried that falling back on every string selection (or item-translation-time) costs more than one time on the set_locale() call

I think both are needed, but for different purposes. The former, at locale-selecton-time, is only really important so that the TUI shows the selected/resolved locale rather than the requested locale (so after checking if the locale exists, splitting etc).  The latter, at item-translation-time, is needed to allow partial translations (a subset of items with translations for a given language).

> Aesthetically I'm curious why this is in an "inner" function, but it's not that important.

It was just so I can add tests again the pure part of the function.  I've reworked/renamed it and added tests. Hopefully the test cases make the various locale-selecton-time fallback logic clearer.